### PR TITLE
daimo-userop package

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
some todos, some clean up left, but things work end to end

https://github.com/daimo-eth/daimo/assets/6984346/9086c4b0-0d99-4388-a0d8-a785b9140adb

### CODE REVIEW

File: packages/daimo-userop/dist/config.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (nalinbhardwaj) [inline packages/daimo-userop/dist/config.json:1]: todo here: these dist folders are the same as https://github.com/daimo-eth/daimo/pull/24#discussion_r1225762167, just need to wrap them into some build pipeline

File: packages/daimo-userop/package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "daimo-userop",
+  "version": "1.0.0",
+  "description": "userop.js for daimo account contracts",
+  "main": "dist/src/index.js",
+  "types": "dist/src/index.d.ts",
+  "scripts": {
+    "build": "tsc",
+    "test": "jest"
```
Comment (nalinbhardwaj) [inline packages/daimo-userop/package.json:9]: todo: needs lint

File: packages/daimo-userop/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        })
+      ),
+      {
+        dryRun: this.dryRun,
+        onBuild: (op) => console.log("[OP] Signed UserOperation:", op),
+      }
+    );
+    console.log(`UserOpHash: ${res.userOpHash}`);
+
+    const ev = await res.wait(); // TODO: use getUserOperationStatus?
```
Comment (nalinbhardwaj) [inline packages/daimo-userop/src/index.ts:117]: this todo can probably be punted for a future PR

File: packages/daimo-userop/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  private async parseErc20Amount(
+    amount: `${number}`,
+    tokenAddress: `0x${string}`
+  ): Promise<bigint> {
+    const erc20 = getContract({
+      abi: Contracts.erc20ABI,
+      address: tokenAddress,
+      publicClient,
+    });
+    const decimals = await erc20.read.decimals(); // TODO: Just hardcode for performance
```
Comment (nalinbhardwaj) [inline packages/daimo-userop/src/index.ts:89]: todo: should just do this for efficiency, txes are pretty slow atm.
can probably also cache some stuff like `getSenderAddress` and warm the cache in app in background before the user even clicks anything

File: packages/expo-enclave/example/App.tsx [hunk 5]
```diff
@@ -22,6 +31,34 @@ export default function App() {
     setBiometricSecurityLevel(await ExpoEnclave.getBiometricSecurityLevel());
   })();
 
+  const testTx = async () => {
+    const accName = "testdaimo";
+    const derPublicKey = (await ExpoEnclave.fetchPublicKey(accName)) as string;
+    const signer: SigningCallback = async (message) => {
+      // hex version of '\x19Ethereum Signed Message:\n32'
+      const hexifiedPrefix =
+        "19457468657265756d205369676e6564204d6573736167653a0a3332";
```
Comment (nalinbhardwaj) [inline packages/expo-enclave/example/App.tsx:40]: todo: this code block can/should probably just be wrapped into the middleware callback in the daimo-userop package for cleanliness

File: packages/expo-enclave/example/App.tsx [hunk 6]
```diff
@@ -22,6 +31,34 @@ export default function App() {
...
+      const signature = await ExpoEnclave.sign(
+        accName,
+        hexifiedPrefix + message,
+        biometricPromptCopy
+      );
+      return signature;
+    };
+    const account = await DaimoAccount.init(derPublicKey, signer, false);
+    console.log(
+      "account, give it some eth + usdc magically:",
```
Comment (nalinbhardwaj) [inline packages/expo-enclave/example/App.tsx:50]: todo: I guess this is fixed magically in the main app so can ignore

File: packages/expo-enclave/example/package.json [hunk 7]
```diff
@@ -9,6 +9,8 @@
     "web": "expo start --web"
   },
   "dependencies": {
+    "daimo-contract-types": "file:daimo-contract-types-1.0.0.tgz",
+    "daimo-userop": "file:daimo-userop-1.0.0.tgz",
```
Comment (nalinbhardwaj) [inline packages/expo-enclave/example/package.json:13]: using tgz here is very weird, apparently react native is unable to take as input local dependencies (stuff with `file:..`), see https://github.com/facebook/react-native/issues/29977 and this seems like one of the suggested fixes. There's probably a better fix by using workspaces or something as in one of the comments above, todo here to investigate/fix

**Review by nalinbhardwaj — COMMENTED** [review]
leaving my TODOs inline so you're not blocked on this @dcposch

File: packages/daimo-userop/README.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Daimo Userop
+
+Simple, opinionated wrapper over a `userop.js` Builder for Daimo P256 contracts. Works in React Native and easy to integrate with P256/external signers. See example app in expo-enclave for usage.
```
Comment (dcposch) [inline packages/daimo-userop/README.md:3]: 🚀 🚀

File: packages/daimo-userop/dist/src/daimoAccountBuilder.d.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import { UserOperationBuilder } from "userop";
+import { UserOperationMiddlewareFn } from "userop";
+import { SigningCallback } from "./util";
+import "@ethersproject/shims";
+import "text-encoding-polyfill";
+export declare class DaimoAccountBuilder extends UserOperationBuilder {
```
Comment (dcposch) [inline packages/daimo-userop/dist/src/daimoAccountBuilder.d.ts:6]: recommend `.gitignore`ing `dist`

Comment (nalinbhardwaj) [reply packages/daimo-userop/dist/src/daimoAccountBuilder.d.ts:6]: would like to, yeah, see https://github.com/daimo-eth/daimo/pull/25#discussion_r1225900274

Comment (dcposch) [reply packages/daimo-userop/dist/src/daimoAccountBuilder.d.ts:6]: i see, makes sense!

**Comment by dcposch** [comment]
combined into #26

**Pull request closed without merging**